### ISSUE TITLE
TypedDict not working with dict variable

### ISSUE THREAD

**Issue by nihaals** [microsoft/pyright#715]
**Describe the bug**
TypedDict validation doesn't seem to work when setting a dict to a variable.

**To Reproduce**
```py
from typing import TypedDict


class Test(TypedDict):
    name: str


class TestClass:
    def __init__(self, name: str) -> None:
        self.name = name

    def as_dict(self) -> Test:
        out = {
            'name': self.name,
        }
        return out
```
2. Notice `out` is given `Expression of type "dict[str, str]" cannot be assigned to return type "Test"
  "dict[str, str]" is incompatible with "Test"`

```py
from typing import TypedDict


class Test(TypedDict):
    name: str


class TestClass:
    def __init__(self, name: str) -> None:
        self.name = name

    def as_dict(self) -> Test:
        return {
            'name': self.name,
        }
```
4. Notice lack of error (intended)

**Expected behavior**
It works

**VS Code extension or command-line**
VS Code Extension 1.1.41

**Additional context**
I know you wouldn't just assign to a variable and return it, this is part of a more complex case with an if statement due to an optional key and this is a MWE I found. This also happens with MyPy which is confusing. Type hinting `out: Test` solves the problem. Not sure if this is an unintuitive feature part of the spec and therefore intentional or a missing feature.

**Comment by erictraut** [comment]
This is the correct behavior. A dict is not compatible with a TypedDict, so an attempt to assign a variable of type dict to a variable of type TypedDict will result in an error.

In your first example above, you have assigned a dictionary literal to variable `out`, but you haven't provided any type annotation for `out`. That forces the type checker to infer its type from the assignment. In this case, it will infer that the type of `out` is `Dict[str, str]`. Then you are attempting to return out, which is incompatible with the declared return type of `Test`. It is correct for an error to be generated in this case.

In the second example, you are directly returning a dictionary literal, and the "expected type" of a return expression in this context is Test. The type checker doesn't need to infer any types here; it simply verifies that the dictionary literal expression conforms to type `Test`. In this case, it is compatible, so no error is generated.

The correct way to handle this in your first example is to provide a type annotation for `out`.
```
out: Test = { 'name': self.name }
```

**Comment by nihaals** [comment]
Thanks for the explanation, after reading the PEP I understand that you should explicitly set a type to a TypedDict or you can assume in the case of parameters which makes sense as setting a type has other impacts
